64_MergeSort/main.cpp

```diff
@@ -0,0 +1,368 @@
+			}
+			else
+			{
+				cmdbuf->bindDescriptorSets(nbl::asset::EPBP_COMPUTE, m_pipelineLayout.get(), 0, 1, &m_descriptorSet2.get());
+			}
+
+			struct SortingPhaseData
+			{
+				uint32_t numElementsPerArray;
+				uint32_t bufferLength;
+			};
+
+			const SortingPhaseData pushConstantData = {
+				.numElementsPerArray = (uint32_t)pow(2, phaseIndex - 1),
+				.bufferLength = static_cast<uint32_t>(NumberOfElementsToSort),
+			};
+
+			cmdbuf->pushConstants(m_pipelineLayout.get(), nbl::asset::IShader::ESS_COMPUTE, 0u, sizeof(pushConstantData), &pushConstantData);
+
+			cmdbuf->dispatch((uint32_t)ceilf(NumberOfElementsToSort / powf(2.0f, phaseIndex)), 1, 1);
+			cmdbuf->endDebugMarker();
+
+			cmdbuf->end();
+
+			{
+				// queues are inherent parts of the device, ergo not refcounted (you refcount the device instead)
+				IQueue* queue = m_device->getQueue(m_computeQueueFamily, 0);
+
+				IQueue::SSubmitInfo submitInfos[1] = {};
+				// The IGPUCommandBuffer is the only object whose usage does not get automagically tracked internally, you're responsible for holding onto it as long as the GPU needs it.
+				// So this is why our commandbuffer, even though its transient lives in the scope equal or above the place where we wait for the submission to be signalled as complete.
+				const IQueue::SSubmitInfo::SCommandBufferInfo cmdbufs[] = { {.cmdbuf = cmdbuf.get()} };
+				submitInfos[0].commandBuffers = cmdbufs;
+
+				const IQueue::SSubmitInfo::SSemaphoreInfo signals[] = { {.semaphore = m_phaseSemaphore.get(),.value = ++monotonicallyIncreasingCounter,.stageMask = asset::PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT} };
+				submitInfos[0].signalSemaphores = signals;
+
+				queue->startCapture();
+				queue->submit(submitInfos);
+				queue->endCapture();
+
+			}
+			const ISemaphore::SWaitInfo wait_infos[] = { {
+		.semaphore = m_phaseSemaphore.get(),
+		.value = monotonicallyIncreasingCounter,
+	}, };
+
+			m_device->blockForSemaphores(wait_infos);
+		}
+
+		// Now, get the current pointer to output buffer (the output may be in either input or output buffer).
+		// Perform merge sort on the CPU to test whether GPU compute version computed the result correctly.
+		auto outputBufferData = reinterpret_cast<const int32_t*>(m_outputBufferAllocation.memory->getMappedPointer());
+
+		if (numberOfPhases % 2 == 0)
+		{
+			outputBufferData = reinterpret_cast<const int32_t*>(m_inputBufferAllocation.memory->getMappedPointer());
+		}
+
+		// Sorted input (which should match the output buffer).
+		std::sort(inputBufferData.begin(), inputBufferData.end());
+
+		for (auto i = 0; i < NumberOfElementsToSort; i++)
+		{
+			printf("output buffer -> %d input buffer -> %d\n", outputBufferData[i], inputBufferData[i]);
+			if (outputBufferData[i] != inputBufferData[i])
+			{
+				return logFail("%d != %d\n", outputBufferData[i], inputBufferData[i]);
+			}
+		}
+
+		printf("Merge sort succesfull!\n");
```

Comment:
use the `m_logger` or whatever you base app has, then all messagess go to same output

Reply:
Done, will stop using printf in the future and use m_logger.